test/test_framework.py

```diff
@@ -1220,91 +1272,99 @@ class CustomObject:
             assert str(e) == \
                 "attribute 'foo' cannot be a CustomObject: must be int/float/dict/list/etc"
         else:
-            self.fail('AttributeError not raised')
+            pytest.fail('AttributeError not raised')
 
         framework.commit()
 
-    def test_mutable_types(self):
+    def test_mutable_types(self, request: pytest.FixtureRequest):
         # Test and validation functions in a list of tuples.
         # Assignment and keywords like del are not supported in lambdas
         #  so functions are used instead.
+        def _assert_equal(a: typing.Any, b: typing.Any):
```

Comment:
I couldn't move this big test into parametrize because you can't define functions in parametrize. I kept it as it is, but changed the unittest `self.assert` to pytest `assert` style using Ben's suggestion from the previous preview PR.

Reply:
I think it's fine to leave like it is for now, and at some point we should refactor this specifically so that it's clearer and doesn't have all the lambdas.

Reply:
I think it's fine to leave like it is for now, and at some point we should refactor this specifically so that it's clearer and doesn't have all the lambdas.